### ISSUE TITLE
R Code

### ISSUE THREAD

**Issue by Rajk7513** [Rajk7513/ruHealthy#1]
// We have analysed the dataset to find correlation between the other variables and user feedback. This can help us build a model to analyse the factors that correlate to the user providing positive or negative feedback. The trends that we observe relate certain categories of people to negative and positive feedbacks however overall positive feedbacks are much more common than negative ones. To do this we have plotted the positive vs negative feedback ratio of these categories. The biggest trend that we observed was that people whose output status was "No" were much more likely to provide negative feedbacks. Other trends we observed were that Students and No income people( who had a huge overlap) tended to provide positive feedbacks and married people tended to give negative feedbacks. Another slightly less significant trend we discovered was that family sizes between 2 and 4 tended to give more positive feedbacks. In Order to increase the accuracy of the rpart model we set the maximum depth of the decision tree to 4, which prevents overfitting and leads to generally accurate model. we achieved an accuracy of [92%.``]

# Load required libraries
library(rpart)
library(rpart.plot)

data <- read.csv("onlinefoods.csv")
summary(data)

#Numerical variables - Age, Family.size, latitude, longitude, Pin.code
#Character variable - Gender, Martial.Status, Occupation, Monthly.Income, EEducational.Qualifications
#Output, Feedback, X

#combining all the variables

num_var <- cbind.data.frame(data)
summary(data)

# Summary statistics for numerical variables
summary(data[, c("Age", "Family.size", "latitude", "longitude", "Pin.code")])

# Train/test split (assuming 70% train and 30% test)
set.seed(123) # for reproducibility
train_index <- sample(1:nrow(data), 0.7 * nrow(data))
train_data <- data[train_index, ]
test_data <- data[-train_index, ]

minBuck <-16
minS <- 15
cup <- 0.01
dep <- 4

# Train the rpart model
model <- rpart(Feedback ~ Age + Family.size + latitude + longitude + Gender + 
               Marital.Status + Occupation + Monthly.Income + Output +
               Educational.Qualifications, 
               data = train_data, method = "class", control = rpart.control(minsplit = minS, minbucket = minBuck, cp = cup, maxdepth = dep))

# Predict on test data
predictions <- predict(model, newdata = test_data, type = "class")

# Confusion matrix
conf_matrix <- table(predictions, test_data$Output)
print(conf_matrix)

# Load required libraries
library(e1071)  # for Naive Bayes classification

# Train Naive Bayes model
nb_model <- naiveBayes(Feedback ~ Output + Age + Family.size + latitude + longitude + Gender + Occupation + Monthly.Income + Educational.Qualifications, data = train_data)

# Predict on training data
nb_predictions <- predict(nb_model, newdata = test_data)

# Confusion matrix for Naive Bayes
nb_conf_matrix <- table(nb_predictions, test_data$Output)
print("Confusion Matrix for Naive Bayes:")
print(nb_conf_matrix)

# Calculate accuracy for Naive Bayes
nb_accuracy <- sum(diag(nb_conf_matrix)) / sum(nb_conf_matrix)
print(paste("Accuracy of Naive Bayes:", nb_accuracy))

# Compare with rpart accuracy
rpart_accuracy <- sum(diag(conf_matrix)) / sum(conf_matrix)
print(paste("Accuracy of rpart:", rpart_accuracy))
